lib/ruby_indexer/lib/ruby_indexer/index.rb

```diff
@@ -434,22 +452,147 @@ def handle_change(indexable)
+        # `RubyLsp::Interface::Location` is allowed. For these cases, we need some way to realize that the
+        # `RubyLsp::Interface` part is an alias, that has to be resolved
+        entries = direct_or_aliased_constant("#{namespace}::#{name}", seen_names)
+        return entries if entries
+      end
+
+      nil
+    end
+
+    sig do
+      params(
+        name: String,
+        nesting: T::Array[String],
+        seen_names: T::Array[String],
+      ).returns(T.nilable(T::Array[Entry]))
+    end
+    def lookup_ancestor_chain(name, nesting, seen_names)
+      *nesting_parts, constant_name = build_non_redundant_full_name(name, nesting).split("::")
+      namespace_entries = resolve(T.must(nesting_parts).join("::"), [], seen_names)
+      return unless namespace_entries
+
+      ancestors = T.must(nesting_parts).empty? ? [] : linearized_ancestors_of(T.must(namespace_entries.first).name)
+
+      ancestors.each do |ancestor_name|
+        entries = direct_or_aliased_constant("#{ancestor_name}::#{constant_name}", seen_names)
+        return entries if entries
+      end
+
+      nil
+    rescue NonExistingNamespaceError
+      nil
+    end
+
+    # Removes redudancy from a constant reference's full name. For example, if we find a reference to `A::B::Foo` inside
+    # of the ["A", "B"] nesting, then we should not concatenate the nesting with the name or else we'll end up with
+    # `A::B::A::B::Foo`. This method will remove any redundant parts from the final name based on the reference and the
+    # nesting
+    sig { params(name: String, nesting: T::Array[String]).returns(String) }
+    def build_non_redundant_full_name(name, nesting)
+      return name if nesting.empty?
+
+      namespace = nesting.join("::")
+
+      # If the name is not qualified, we can just concatenate the nesting and the name
+      return "#{namespace}::#{name}" unless name.include?("::")
+
+      name_parts = name.split("::")
+
+      # Find the first part of the name that is not in the nesting
+      index = name_parts.index { |part| !nesting.include?(part) }
+
+      if index.nil?
+        # All parts of the nesting are redundant because they are already present in the name. We can return the name
+        # directly
+        name
+      elsif index == 0
+        # No parts of the nesting are in the name, we can concatenate the namespace and the name
+        "#{namespace}::#{name}"
+      else
+        # The name includes some parts of the nesting. We need to remove the redundant parts
+        "#{namespace}::#{T.must(name_parts[index..-1]).join("::")}"
+      end
+    end
+
+    sig { params(full_name: String, seen_names: T::Array[String]).returns(T.nilable(T::Array[Entry])) }
+    def direct_or_aliased_constant(full_name, seen_names)
+      entries = @entries[full_name] || @entries[follow_aliased_namespace(full_name)]
+      entries&.map { |e| e.is_a?(Entry::UnresolvedAlias) ? resolve_alias(e, seen_names) : e }
```

Comment:
I was thinking of extracting just this part so it can be better reused, example lines 189 and 509

Reply:
I extracted the specific usage of lines 189 and 509 instead. This method is used more often than the usage of lines 189 and 509, it is used 5 times.